src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
 		NULL);
 }
 
+#define STACK_ALLOC 256
+
+// The maximum number of type parameters for a BulkTypeValue instance
+// Aligned with coreCLR StackSArray<ULONGLONG> rgTypeParameters
+#define K_SIZE_OF_TYPE_PARAMETER_ARRAY ((uint32_t)(STACK_ALLOC / sizeof (intptr_t)))
+
+// !!!!!!! NOTE !!!!!!!!
+// The flags must match those in the ETW manifest exactly
+// !!!!!!! NOTE !!!!!!!!
+
+typedef enum {
+	K_ETW_TYPE_FLAGS_DELEGATE = 0x1,
+	K_ETW_TYPE_FLAGS_FINALIZABLE = 0x2,
+	K_ETW_TYPE_FLAGS_EXTERNALLY_IMPLEMENTED_COM_OBJECT = 0x4,
+	K_ETW_TYPE_FLAGS_ARRAY = 0x8,
+
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_MASK = 0x3F00,
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_SHIFT = 8,
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_MAX = K_ETW_TYPE_FLAGS_ARRAY_RANK_MASK >> K_ETW_TYPE_FLAGS_ARRAY_RANK_SHIFT
+} EtwTypeFlags;
+
+// This only contains the fixed-size data at the top of each struct in
+// the bulk type event.  These fields must still match exactly the initial
+// fields of the struct described in the manifest.
+typedef struct _EventStructBulkTypeFixedSizedData {
+	uint64_t type_id;
+	uint64_t module_id;
+	uint32_t type_name_id;
+	uint32_t flags;
+	uint8_t cor_element_type;
+} EventStructBulkTypeFixedSizedData;
+
+// Represents one instance of the Value struct inside a single BulkType event
+typedef struct _BulkTypeValue {
+	EventStructBulkTypeFixedSizedData fixed_sized_data;
+	uint32_t c_type_parameters;
+	intptr_t rg_type_parameters [K_SIZE_OF_TYPE_PARAMETER_ARRAY];
+	MonoType *rg_mono_type_parameters [K_SIZE_OF_TYPE_PARAMETER_ARRAY];
+	ep_char8_t *s_name; // Currently should only be NULL, TODO if we want to provide the name in the BulkTypeEvent data, figure out memory management to use
+} BulkTypeValue;
+
+static
+void
+ep_rt_bulk_type_value_clear (BulkTypeValue *bulk_type_value);
+
+static
+int
+ep_rt_mono_get_byte_count_in_event (BulkTypeValue *bulk_type_value);
+
+static
+BulkTypeEventLogger*
+ep_rt_bulk_type_event_logger_init (void);
+
+static
+void
+ep_rt_bulk_type_event_logger_fini (BulkTypeEventLogger *p_type_logger);
+
+static
+int
+write_event_buffer (
+	const uint8_t *val,
+	size_t size,
+	char *buf_start,
+	char **buf_next);
+
+static
+int
+write_event_buffer_int8 (
+	int8_t val,
+	char *buf_start,
+	char **buf_next);
+
+static
+int
+write_event_buffer_int16 (
+	int16_t val,
+	char *buf_start,
+	char **buf_next);
+
+static
+int
+write_event_buffer_int32 (
+	int32_t val,
+	char *buf_start,
+	char **buf_next);
+
+static
+int
+write_event_buffer_int64 (
+	int64_t val,
+	char *buf_start,
+	char **buf_next);
+
+static
+intptr_t
+get_typeid_for_type (MonoType *t);
+
+static
+intptr_t
+get_typeid_for_class (MonoClass *c);
+
+// Clear out BulkTypeValue before filling it out (array elements can get reused if there
+// are enough types that we need to flush to multiple events).
+static
+void
+ep_rt_bulk_type_value_clear (BulkTypeValue *bulk_type_value)
+{
+	bulk_type_value->fixed_sized_data.type_id = 0;
```

Comment:
nit: you could do a memset on the sizeof(BulkTypeValue) instead of setting all individual fields to 0. That would also be more resistant if new fields are added, since they would be automatically cleared.